<a href="https://colab.research.google.com/github/Mizoroki-Heck/Statistics/blob/main/LinearRegressions/%D0%9B%D0%B8%D0%BD%D0%B5%D0%B9%D0%BD%D0%B0%D1%8F%20%D1%80%D0%B5%D0%B3%D1%80%D0%B5%D1%81%D1%81%D0%B8%D1%8F%20%D0%BE%D0%BF%D0%BE%D1%81%D1%81%D1%83%D0%BC%D0%BE%D0%B2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.model_selection import train_test_split
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.metrics import mean_squared_error

Сперва без категориальных переменных

In [ ]:
data1 = pd.read_csv('/content/sample_data/possum.csv')
data1

case  site    Pop sex  age  hdlngth  skullw  totlngth  taill  footlgth  \
0       1     1    Vic   m  8.0     94.1    60.4      89.0   36.0      74.5   
1       2     1    Vic   f  6.0     92.5    57.6      91.5   36.5      72.5   
2       3     1    Vic   f  6.0     94.0    60.0      95.5   39.0      75.4   
3       4     1    Vic   f  6.0     93.2    57.1      92.0   38.0      76.1   
4       5     1    Vic   f  2.0     91.5    56.3      85.5   36.0      71.0   
..    ...   ...    ...  ..  ...      ...     ...       ...    ...       ...   
99    100     7  other   m  1.0     89.5    56.0      81.5   36.5      66.0   
100   101     7  other   m  1.0     88.6    54.7      82.5   39.0      64.4   
101   102     7  other   f  6.0     92.4    55.0      89.0   38.0      63.5   
102   103     7  other   m  4.0     91.5    55.2      82.5   36.5      62.9   
103   104     7  other   f  3.0     93.6    59.9      89.0   40.0      67.6   

     earconch   eye  chest  belly  
0        54.5  15.2   28.0   36.0  
1        51.2  16.0   28.5   33.0  
2        51.9  15.5   30.0   34.0  
3        52.2  15.2   28.0   34.0  
4        53.2  15.1   28.5   33.0  
..        ...   ...    ...    ...  
99       46.8  14.8   23.0   27.0  
100      48.0  14.0   25.0   33.0  
101      45.4  13.0   25.0   30.0  
102      45.9  15.4   25.0   29.0  
103      46.0  14.8   28.5   33.5  

[104 rows x 14 columns]

In [ ]:
data1 = data1.drop(['case', 'site', 'Pop', 'sex'], axis=1)
data1.dropna(inplace=True)
data1

age  hdlngth  skullw  totlngth  taill  footlgth  earconch   eye  chest  \
0    8.0     94.1    60.4      89.0   36.0      74.5      54.5  15.2   28.0   
1    6.0     92.5    57.6      91.5   36.5      72.5      51.2  16.0   28.5   
2    6.0     94.0    60.0      95.5   39.0      75.4      51.9  15.5   30.0   
3    6.0     93.2    57.1      92.0   38.0      76.1      52.2  15.2   28.0   
4    2.0     91.5    56.3      85.5   36.0      71.0      53.2  15.1   28.5   
..   ...      ...     ...       ...    ...       ...       ...   ...    ...   
99   1.0     89.5    56.0      81.5   36.5      66.0      46.8  14.8   23.0   
100  1.0     88.6    54.7      82.5   39.0      64.4      48.0  14.0   25.0   
101  6.0     92.4    55.0      89.0   38.0      63.5      45.4  13.0   25.0   
102  4.0     91.5    55.2      82.5   36.5      62.9      45.9  15.4   25.0   
103  3.0     93.6    59.9      89.0   40.0      67.6      46.0  14.8   28.5   

     belly  
0     36.0  
1     33.0  
2     34.0  
3     34.0  
4     33.0  
..     ...  
99    27.0  
100   33.0  
101   30.0  
102   29.0  
103   33.5  

[101 rows x 10 columns]

In [ ]:
X = data1.drop(['totlngth'], axis=1)
y = data1['totlngth']
X = (X - X.mean(axis=0)) / (X.std(axis=0))
X_constant = sm.add_constant(X)
X_train, X_test, y_train, y_test = train_test_split(X_constant, y, train_size=0.7, test_size=0.3, random_state = 42)

In [ ]:
results1 = sm.OLS(y_train, X_train).fit()
results1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               totlngth   R-squared:                       0.739
Model:                            OLS   Adj. R-squared:                  0.700
Method:                 Least Squares   F-statistic:                     18.92
Date:                Sat, 16 Dec 2023   Prob (F-statistic):           1.78e-14
Time:                        14:04:27   Log-Likelihood:                -152.19
No. Observations:                  70   AIC:                             324.4
Df Residuals:                      60   BIC:                             346.9
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         87.3922      0.278    314.229      0.000      86.836      87.949
age            0.0568      0.312      0.182      0.856      -0.568       0.682
hdlngth        0.9098      0.443      2.052      0.045       0.023       1.797
skullw         0.0872      0.395      0.221      0.826      -0.703       0.877
taill          2.3252      0.339      6.858      0.000       1.647       3.003
footlgth       1.3138      0.586      2.244      0.029       0.143       2.485
earconch       0.2545      0.575      0.443      0.659      -0.895       1.404
eye            0.0794      0.314      0.253      0.801      -0.550       0.708
chest          0.5597      0.417      1.342      0.185      -0.275       1.394
belly          0.1058      0.366      0.289      0.774      -0.626       0.838
==============================================================================
Omnibus:                        8.762   Durbin-Watson:                   2.231
Prob(Omnibus):                  0.013   Jarque-Bera (JB):                8.338
Skew:                          -0.806   Prob(JB):                       0.0155
Kurtosis:                       3.512   Cond. No.                         5.51
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Посчитаем VIF

In [ ]:
vif = pd.DataFrame()
vif['VIF'] = [variance_inflation_factor(X_train.values,i) for i in range(X_train.shape[1])]
vif['variable'] = X_train.columns
vif

VIF  variable
0  1.024775     const
1  1.262334       age
2  2.528610   hdlngth
3  2.261395    skullw
4  1.510153     taill
5  4.437057  footlgth
6  4.298933  earconch
7  1.286140       eye
8  2.259688     chest
9  2.021025     belly

Значение 1 указывает на отсутствие корреляции между данной независимой переменной и любыми другими независимыми переменными в модели.   

Значение от 1 до 5 указывает на умеренную корреляцию между данной объясняющей переменной и другими независимыми переменными в модели, но часто она недостаточно серьезна, чтобы требовать внимания.

Видим, что коэфициент детерминации выше 0.7, чем ближе значение коэффициента к 1, тем сильнее зависимость.
Коэфициент кореляции это корень из R^2 и он равен 0.85965109201, что говорит о сильной связи.

In [ ]:
predict_train = results1.predict(X_train)
predict_test = results1.predict(X_test)

mse_train = mean_squared_error(y_train, predict_train)
mse_test = mean_squared_error(y_test, predict_test)

print(f'MSE на тренировочном наборе: {mse_train:.4f}')
print(f'MSE на тестовом наборе: {mse_test:.4f}')

MSE на тренировочном наборе: 4.5287
MSE на тестовом наборе: 5.4400


Теперь попробуем с категориальными данными

In [ ]:
data2 = pd.read_csv('/content/sample_data/possum.csv')
data2

case  site    Pop sex  age  hdlngth  skullw  totlngth  taill  footlgth  \
0       1     1    Vic   m  8.0     94.1    60.4      89.0   36.0      74.5   
1       2     1    Vic   f  6.0     92.5    57.6      91.5   36.5      72.5   
2       3     1    Vic   f  6.0     94.0    60.0      95.5   39.0      75.4   
3       4     1    Vic   f  6.0     93.2    57.1      92.0   38.0      76.1   
4       5     1    Vic   f  2.0     91.5    56.3      85.5   36.0      71.0   
..    ...   ...    ...  ..  ...      ...     ...       ...    ...       ...   
99    100     7  other   m  1.0     89.5    56.0      81.5   36.5      66.0   
100   101     7  other   m  1.0     88.6    54.7      82.5   39.0      64.4   
101   102     7  other   f  6.0     92.4    55.0      89.0   38.0      63.5   
102   103     7  other   m  4.0     91.5    55.2      82.5   36.5      62.9   
103   104     7  other   f  3.0     93.6    59.9      89.0   40.0      67.6   

     earconch   eye  chest  belly  
0        54.5  15.2   28.0   36.0  
1        51.2  16.0   28.5   33.0  
2        51.9  15.5   30.0   34.0  
3        52.2  15.2   28.0   34.0  
4        53.2  15.1   28.5   33.0  
..        ...   ...    ...    ...  
99       46.8  14.8   23.0   27.0  
100      48.0  14.0   25.0   33.0  
101      45.4  13.0   25.0   30.0  
102      45.9  15.4   25.0   29.0  
103      46.0  14.8   28.5   33.5  

[104 rows x 14 columns]

In [ ]:
data2.sex = data2.sex.map({'m': 1, 'f': 0})
data2.Pop = data2.Pop.map({'Vic': 1, 'other': 0})
data2 = data2.join(pd.get_dummies(data2.site, prefix='site'))
data2.drop(['site', 'case'], axis=1, inplace=True )
data2.dropna(inplace=True)

In [ ]:
data2

Pop  sex  age  hdlngth  skullw  totlngth  taill  footlgth  earconch  \
0      1    1  8.0     94.1    60.4      89.0   36.0      74.5      54.5   
1      1    0  6.0     92.5    57.6      91.5   36.5      72.5      51.2   
2      1    0  6.0     94.0    60.0      95.5   39.0      75.4      51.9   
3      1    0  6.0     93.2    57.1      92.0   38.0      76.1      52.2   
4      1    0  2.0     91.5    56.3      85.5   36.0      71.0      53.2   
..   ...  ...  ...      ...     ...       ...    ...       ...       ...   
99     0    1  1.0     89.5    56.0      81.5   36.5      66.0      46.8   
100    0    1  1.0     88.6    54.7      82.5   39.0      64.4      48.0   
101    0    0  6.0     92.4    55.0      89.0   38.0      63.5      45.4   
102    0    1  4.0     91.5    55.2      82.5   36.5      62.9      45.9   
103    0    0  3.0     93.6    59.9      89.0   40.0      67.6      46.0   

      eye  chest  belly  site_1  site_2  site_3  site_4  site_5  site_6  \
0    15.2   28.0   36.0       1       0       0       0       0       0   
1    16.0   28.5   33.0       1       0       0       0       0       0   
2    15.5   30.0   34.0       1       0       0       0       0       0   
3    15.2   28.0   34.0       1       0       0       0       0       0   
4    15.1   28.5   33.0       1       0       0       0       0       0   
..    ...    ...    ...     ...     ...     ...     ...     ...     ...   
99   14.8   23.0   27.0       0       0       0       0       0       0   
100  14.0   25.0   33.0       0       0       0       0       0       0   
101  13.0   25.0   30.0       0       0       0       0       0       0   
102  15.4   25.0   29.0       0       0       0       0       0       0   
103  14.8   28.5   33.5       0       0       0       0       0       0   

     site_7  
0         0  
1         0  
2         0  
3         0  
4         0  
..      ...  
99        1  
100       1  
101       1  
102       1  
103       1  

[101 rows x 19 columns]

In [ ]:
X = data2.drop(['totlngth'], axis=1)
y = data2['totlngth']
X = (X - X.mean(axis=0)) / (X.std(axis=0))
X_constant = sm.add_constant(X)
X_train, X_test, y_train, y_test = train_test_split(X_constant, y, train_size=0.7, test_size=0.3, random_state = 42)

In [ ]:
results2 = sm.OLS(y_train, X_train).fit()
results2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               totlngth   R-squared:                       0.828
Model:                            OLS   Adj. R-squared:                  0.777
Method:                 Least Squares   F-statistic:                     15.99
Date:                Sat, 16 Dec 2023   Prob (F-statistic):           7.01e-15
Time:                        14:04:27   Log-Likelihood:                -137.57
No. Observations:                  70   AIC:                             309.1
Df Residuals:                      53   BIC:                             347.4
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         87.3667      0.242    361.763      0.000      86.882      87.851
Pop           -0.0828      0.454     -0.182      0.856      -0.992       0.827
sex           -0.2378      0.283     -0.840      0.405      -0.806       0.330
age            0.2079      0.285      0.729      0.469      -0.364       0.780
hdlngth        1.0147      0.420      2.418      0.019       0.173       1.856
skullw         0.3408      0.359      0.950      0.346      -0.379       1.060
taill          2.0956      0.362      5.789      0.000       1.370       2.822
footlgth       1.2890      0.644      2.001      0.051      -0.003       2.581
earconch       0.2352      0.766      0.307      0.760      -1.301       1.772
eye           -0.3248      0.325     -0.999      0.322      -0.977       0.327
chest          0.4400      0.387      1.138      0.260      -0.335       1.215
belly         -0.1168      0.332     -0.351      0.727      -0.784       0.550
site_1         0.4873      0.373      1.305      0.198      -0.262       1.236
site_2        -0.9022      0.296     -3.043      0.004      -1.497      -0.308
site_3         0.4398      0.241      1.824      0.074      -0.044       0.923
site_4        -0.0498      0.250     -0.199      0.843      -0.551       0.451
site_5         0.2453      0.278      0.882      0.382      -0.312       0.803
site_6         0.1292      0.342      0.378      0.707      -0.557       0.816
site_7        -0.4796      0.237     -2.020      0.048      -0.956      -0.003
==============================================================================
Omnibus:                        0.680   Durbin-Watson:                   2.251
Prob(Omnibus):                  0.712   Jarque-Bera (JB):                0.399
Skew:                          -0.183   Prob(JB):                        0.819
Kurtosis:                       3.056   Cond. No.                     9.60e+15
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 3.43e-30. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

В данном случае, видим выше коэфициент детерминации равные 0.828, коэфициент корреляции 0.90994505328, что говорит о сильной связи.


In [ ]:
vif = pd.DataFrame()
vif['VIF'] = [variance_inflation_factor(X_train.values,i) for i in range(X_train.shape[1])]
vif['variable'] = X_train.columns
vif

/usr/local/lib/python3.10/dist-packages/statsmodels/stats/outliers_influence.py:198: RuntimeWarning: divide by zero encountered in double_scalars
  vif = 1. / (1. - r_squared_i)


VIF  variable
0    1.036551     const
1         inf       Pop
2    1.408311       sex
3    1.411174       age
4    3.039177   hdlngth
5    2.503531    skullw
6    2.308703     taill
7    7.206830  footlgth
8   10.251513  earconch
9    1.843591       eye
10   2.603687     chest
11   2.236932     belly
12        inf    site_1
13        inf    site_2
14        inf    site_3
15        inf    site_4
16        inf    site_5
17        inf    site_6
18        inf    site_7

Тут мы видим громадные значения, (inf) это обычно указывает на сильную линейную зависимость между предикторами, есть мультиколлинеарность.

In [ ]:
predict_train = results2.predict(X_train)
predict_test = results2.predict(X_test)

mse_train = mean_squared_error(y_train, predict_train)
mse_test = mean_squared_error(y_test, predict_test)

print(f'MSE на тренировочном наборе: {mse_train:.4f}')
print(f'MSE на тестовом наборе: {mse_test:.4f}')

MSE на тренировочном наборе: 2.9822
MSE на тестовом наборе: 4.7213


Т.К чем MSE ближе к нулю тем лучше, делаем вывод что во втором случае с категориальными признаками, лучше чем без.
Но во втором случае видим, что VIf показывает мультиколлинеарность, что не есть хорошо.